### PR TITLE
tools: scripts: generic script for flashing non Xilinx platforms

### PR DESCRIPTION
**Pull request by AAndrisa**
## Pull Request Description

This script is an adaptation of the maxfla.sh script which now offers support for other two non Xilinx platforms, STM32 and Pico.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: tools/scripts/mcufla.sh [hunk 1]
```diff
@@ -109,10 +123,47 @@ build_prerequisites() {
...
+		build_prerequisites_stm32
+		print_run ~/.stm32/openocd/src/openocd -s ~/.stm32/openocd/tcl \
+			-f interface/stlink-dap.cfg -c "transport select dapdirect_swd; $DAP_SERIAL" \
+			-f "target/$TAR" -c "program $1 verify reset exit"
+	elif [ $PLATFORM == 'pico' ]; then
+		echo "Pico platform detected"
+		print_run openocd -f interface/cmsis-dap.cfg -f "target/$TARGET" \
+			-c "adapter speed 5000; $DAP_SERIAL" -c "program $1 verify reset exit"
+	else
+		echo "Platform was not correct detected"
```
Comment (CiprianRegus) [inline tools/scripts/mcufla.sh:163]: correct -> correctly or just: invalid platform.

File: tools/scripts/mcufla.sh [hunk 2]
```diff
@@ -33,17 +33,18 @@
...
 USAGE=$(cat <<-END
-Flash maxim microcontrollers using Maxim's own openocd.
-Only use .elf files, won't work with .hex files.
+Flash maxim, stm32 and pico microcontrollers, it downloads specific openocd for Maxim and STM32
+platform, for Pico standard openocd is required.
+Can be used only with .elf files.
 
 Dependencies (Debian):
-	autoconf automake make git libudev-dev libtool libusb-1.0-0-dev pkg-config texinfo
+	autoconf automake make git libudev-dev libtool libusb-1.0-0-dev pkg-config texinfo openocd
```
Comment (CiprianRegus) [inline tools/scripts/mcufla.sh:41]: `openocd` is downloaded by this script. I think this list is only meant to include dependencies which have to be installed by the user.

File: tools/scripts/mcufla.sh [hunk 3]
```diff
@@ -33,17 +33,18 @@
...
 
 Usage:
 	maxfla.sh path/to/binary.elf [adapter_serial_number]
 
 Author:
-	darius.berghe@analog.com
+	darius.berghe@analog.com and andreea.andrisan@analog.com
```
Comment (CiprianRegus) [inline tools/scripts/mcufla.sh:47]: Add your name on a new line. There might be other people contributing in the future and it's better formatted that way.

File: tools/scripts/mcufla.sh [hunk 4]
```diff
@@ -109,10 +123,47 @@ build_prerequisites() {
...
+	elif [ $PLATFORM == 'stm32' ]; then
+		echo "STM32 platform detected"
+		TAR="$TARGET"x.cfg
+		build_prerequisites_stm32
+		print_run ~/.stm32/openocd/src/openocd -s ~/.stm32/openocd/tcl \
+			-f interface/stlink-dap.cfg -c "transport select dapdirect_swd; $DAP_SERIAL" \
+			-f "target/$TAR" -c "program $1 verify reset exit"
+	elif [ $PLATFORM == 'pico' ]; then
+		echo "Pico platform detected"
+		print_run openocd -f interface/cmsis-dap.cfg -f "target/$TARGET" \
```
Comment (CiprianRegus) [inline tools/scripts/mcufla.sh:160]: What version of openocd does pico work with. Shouldn't it also be installed similar to the other platforms?

**Review by CiprianRegus — COMMENTED** [review]
The commit message has to be changed to include the maxim platform as well.

**Review by amiclaus — COMMENTED** [review]
please update the `Usage:` also, the script was renamed.

File: tools/scripts/mcufla.sh [hunk 5]
```diff
@@ -109,10 +123,47 @@ build_prerequisites() {
...
+}
 
-print_run ~/.maxflash/openocd/src/openocd -s ~/.maxflash/openocd/tcl \
-	  -c "adapter driver cmsis-dap; transport select swd; $DAP_SERIAL" \
-	  -f "target/$TARGET.cfg"  -c "program $1 verify reset exit"
+configure_specific_platform() {
+	if [ $PLATFORM == 'maxim' ]; then
+		echo "Maxim platform detected"
+		build_prerequisites_maxim
+		print_run ~/.maxflash/openocd/src/openocd -s ~/.maxflash/openocd/tcl \
```
Comment (buha) [inline tools/scripts/mcufla.sh:148]: `maxflash` -> `mcuflash` throughout the file

Comment (buha) [reply tools/scripts/mcufla.sh:148]: bad phrasing, sorry, I meant to say we should use `~/.mcuflash/` folder to store openocd sources and builds

File: tools/scripts/mcufla.sh (hunk 4 above)
Comment (buha) [reply tools/scripts/mcufla.sh:160]: Rpi pico devs managed to upstream their configuration files and so we can use stock openocd.

Apparently it even works with the ST fork, probably they upstreamed it long ago or ST devs rebase often (I didn't check).

**Comment by AAndrisa** [comment]
v2: the path were openocd repositories are cloned is now ~/.mcuflash and every platform has a separate directory in there.
Pico is also set to use openocd from STM32 platform, in this way the openocd  dependency was deleted because will be automatically downloaded by the script.
Usage message and also Authors fields were modified.

**Review by CiprianRegus — APPROVED** [review]
looks good!

**Review by buha — APPROVED** [review]

**Review by SRaus — APPROVED** [review]

**Pull request merged**